src/persist-client/src/internal/state_diff.rs

```diff
@@ -263,21 +263,41 @@ impl<T: Timestamp + Lattice + Codec64> StateDiff<T> {
         batches.chain(rollups)
     }
 
-    pub(crate) fn blob_deletes(&self) -> impl Iterator<Item = HollowBlobRef<T>> {
-        let batches = self
+    pub(crate) fn part_deletes(&self) -> impl Iterator<Item = &RunPart<T>> {
+        // With the introduction of incremental compaction, we
+        // need to be more careful about what we consider "deleted".
+        // If there is a HollowBatch that we replace 2 out of the 4 runs of,
+        // we need to ensure that we only delete the runs that are actually
+        // no longer referenced.
+        let removed = self
             .referenced_batches()
             .filter_map(|spine_diff| match spine_diff {
                 Insert(_) => None,
-                Update(a, _) | Delete(a) => Some(HollowBlobRef::Batch(a)),
+                Update(a, _) | Delete(a) => Some(a.parts.iter().collect::<Vec<_>>()),
             });
-        let rollups = self
-            .rollups
+        let added = self
+            .referenced_batches()
+            .filter_map(|spine_diff| match spine_diff {
+                Insert(a) => Some(a.parts.iter().collect::<Vec<_>>()),
+                Update(_, a) => Some(a.parts.iter().collect::<Vec<_>>()),
+                Delete(_) => None,
+            })
+            .collect::<Vec<_>>();
```

Comment:
Possible to use a set or similar here instead? I think this is quadratic as written, which we usually try to avoid in non-debug-only codepaths, since it tends to catch up to us eventually...